Allow passing $true/$false as parameter to script using -File

SteveL-MSFT opened the pull request:
Using powershell.exe to execute a PowerShell script using `-File` currently provides no way to pass $true/$false as parameter values.  Current behavior is that `-File` accumulates passed parameters as strings only.

Fix is to special case this based on discussion with PS-Committee to support $true/$false as parsed values to parameters.  Switch values is also supported as currently documented syntax doesn't work.

Fix https://github.com/PowerShell/PowerShell/issues/4036

Doc change is https://github.com/PowerShell/PowerShell-Docs/pull/1430

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to CONTRIBUTING.md.

-->

daxian-dbw commented:
The commits has changes in `src/Modules/Shared/Pester` which should not be included. I think you need to run `git submodule update` to resolve it.

SteveL-MSFT commented:
@daxian-dbw fixed

SteveL-MSFT commented:
@lzybkr if you get a chance, can you review this?  thanks

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -939,8 +951,7 @@ private bool ParseFile(string[] args, ref int i, bool noexitSeen)
                     {
-                        Match m = argPattern.Match(arg);
-                        if (m.Success)
+                        if (arg.Contains(":"))
```
Instead of walking the string twice, move the call to IndexOf to before the `if` and test `offset` >= 0 instead.

SteveL-MSFT replied:
good point, will fix

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -949,7 +960,15 @@ private bool ParseFile(string[] args, ref int i, bool noexitSeen)
                             else
                             {
-                                _collectedArgs.Add(new CommandParameter(arg.Substring(0, offset), arg.Substring(offset + 1)));
+                                bool? boolValue = GetBoolValue(arg.Substring(offset + 1));
```
In the common case, you are creating the same string twice with `arg.Substring(offiset + 1)`, once here, and once below.

lzybkr replied:
So this can be made much simpler, something like:
```c#
string argValue = arg.Substring(offset + 1);
bool? boolValue = GetBoolValue(argValue);
_collectedArgs.Add(new CommandParameter(arg.Substring(0, offset), boolValue ?? argValue);
```

SteveL-MSFT replied:
will fix

SteveL-MSFT replied:
null-coalescing operator requires same type on both sides, I can use ternary operator to simplify it instead

SteveL-MSFT replied:
Nevermind, even ternary expects same type

iSazonov replied:
We can use out parameter or return argValue as default value.

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -959,7 +978,15 @@ private bool ParseFile(string[] args, ref int i, bool noexitSeen)
                     else
                     {
-                        _collectedArgs.Add(new CommandParameter(null, arg));
+                        bool? boolValue = GetBoolValue(arg);
```
Again, you can simply a bunch like above with `??`.

SteveL-MSFT replied:
will fix

SteveL-MSFT replied:
same as above

lzybkr replied:
You can still simplify though, something like:
```c#
object argValue = arg.Substring(offset + 1);
 _collectedArgs.Add(new CommandParameter(arg.Substring(0, offset), GetArgPossiblyAsBool(argValue)));
```

And `GetArgPossiblyAsBool` returns `object`.

SteveL-MSFT replied:
Will change

SteveL-MSFT replied:
Doesn't work, can't implicitly convert object to bool/string

lzybkr commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -201,7 +201,37 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
         }
 
-        It "-File should return exit code from script: <Filename>" -TestCases @(
+        It "-File should be able to pass bool values as strings to parameters: <BoolString>" -TestCases @(
```
I don't think these tests reflect the real user scenario.

Passing `$BoolString` is completely different than passing `$true`.

`$BoolString` is getting replaced with `$true`, but `$true` gets replaced with `True`.

SteveL-MSFT replied:
See comment above regarding usage from different shells.

lzybkr replied:
We should have a test that accepts a `string` parameter and you pass `true` and `false` - we want to make sure that when we convert to `bool`, the `bool` gets properly converted back to the correct string.

And come to think about it, there is a problem - we'll lose the case that the user specified.

SteveL-MSFT replied:
Yeah, that's a problem.  I've changed it so it only works with switches and added tests for other cases

lzybkr commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -858,6 +858,19 @@ private bool ParseFile(string[] args, ref int i, bool noexitSeen)
 
+            bool? GetBoolValue(string arg)
+            {
+                if (arg.Equals("$true", StringComparison.OrdinalIgnoreCase))
```
I think this needs to be `True`, not `$true`.

Or maybe it needs to be both `$true` and `True` - because it depends on the shell how `$true` is expanded.

SteveL-MSFT replied:
Isn't the use case:
```
powershell -file foo.ps1 -myswitch:$true
```

Not clear to me when we need `True` as I don't think we want to support:

```
powershell -file foo.ps1 -myswitch:true
```

SteveL-MSFT replied:
Ok, I think I understand what you're getting it by your comment below.  From a non-powershell shell like bash (which I thought would be the primary use case), I would think we want to support PowerShell syntax for passing $true and $false.  From within powershell, I suppose we shouldn't expect the user to escape $true to be just a string.  I can make this change.

SteveL-MSFT committed:
-File accummulates passed parameters as strings which means $true/$false cannot be passed
as a parameter/switch value.
Fix is to special case this based on discussion with PS-Committee.

SteveL-MSFT commented:
@lzybkr feedback addressed

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -939,17 +951,25 @@ private bool ParseFile(string[] args, ref int i, bool noexitSeen)
-                                _collectedArgs.Add(new CommandParameter(arg.Substring(0, offset), arg.Substring(offset + 1)));
+                                string argValue = arg.Substring(offset + 1);
+                                bool? boolValue = GetBoolValue(argValue);
+                                if (boolValue != null)
```
Sorry again -
Can we use `LanguagePrimitives.TryConvertTo` or `LanguagePrimitives.TryConvertArg`?
Can we use TryGetBoolValue pattern?

```c#
                               string argValue1 = arg.Substring(offset + 1); 
                               string argValue0 = arg.Substring(0, offset);
                               if (TryGetBoolValue(argValue1, out bool boolValue))  
                               {  
                                       _collectedArgs.Add(new CommandParameter(argValue0, boolValue));  
                              }  
                               else  
                               {  
                                       _collectedArgs.Add(new CommandParameter(argValue0, argValue1));  
                               }  
```

SteveL-MSFT replied:
Sure

SteveL-MSFT committed:
only support $true/$false strings for switches, added more tests, changed to use Tester-Doer pattern

iSazonov commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -201,31 +201,35 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
+            @{BoolString = '$truE'},
+            @{BoolString = '$falSe'},
+            @{BoolString = 'trUe'},
+            @{BoolString = 'faLse'}
```
Please add a protection comment like "Don't change case!".

SteveL-MSFT replied:
sure

SteveL-MSFT commented:
@lzybkr any other concerns?

ChristopherGLewis commented:
C:\Temp>type test.ps1
Param( [bool] $test )
Echo $test
C:\Temp>pwsh.exe -file test.ps1 -test $false
test.ps1: Cannot process argument transformation on parameter 'test'. Cannot convert value "System.String" to type "System.Boolean". Boolean parameters accept only Boolean values and numbers, such as $True, $False, 1 or 0.

C:\Temp>pwsh.exe -command $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

iSazonov commented:
@ChristopherGLewis See #10838
